### DISCUSSION TITLE
[nodejs] Is it possible to run more than one named rule at a time?

### DISCUSSION THREAD

**Discussion by Madd0g** [cozodb/cozo#51] [Q&A]
I found the example file - https://github.com/cozodb/cozo/blob/main/cozo-lib-nodejs/example.js

```js

    db.registerNamedRule('Pipipy', 1, async (inputs, options) => {
        console.log(`rule inputs: ${JSON.stringify(inputs)} ${JSON.stringify(options)}`)
        await sleep(1000);
        return inputs[0].map((row) => [row[0] * options.mul])
    })

    try {
        let r = await db.run(`
        rel[] <- [[1],[2]]
        
        ?[a] <~ Pipipy(rel[], mul: 3)
        `);
        console.log(r);
    } catch (e) {
        console.error(e.display);
    }

```

is there a proper way of passing `a` back again to `Pipify`?

I tried the following and different variations, like using a semicolon instead of the comma:

```
?[a] <~ Pipipy(rel[], mul: 3), [b] <~ Pipipy(a[], mul: 5)
```

**Comment by zh217** [comment]
You can chain queries https://docs.cozodb.org/en/latest/stored.html#chaining-queries

If query chaining is not enough for you, you can try multi-statement transaction https://docs.cozodb.org/en/latest/stored.html#multi-statement-transaction , NodeJS examples: https://github.com/cozodb/cozo/blob/v0.5.1/cozo-lib-nodejs/example.js#L45-L54 .

Hope this answers your question.

**Reply by Madd0g** [reply]
thanks, is something like this supposed to work? I saw something similar where in the node example where one `tx.run` was using `a` without defining it.

```js
try {
  const tx = db.multiTransact(true);
  await tx.run(`rel[] <- [[1],[2]]
    
    ?[a] <~ Pipipy(rel[], mul: 3)
    `);

  await tx.run("?[b] <~ Pipipy(a[], mul: 2)");

  console.log(await tx.commit());
} catch (e) {
  console.error(e.display);
}

```

My initial goal was to figure out how to return both a and b, but I couldn't make it that far
